wcfsetup/install/files/lib/system/bbcode/MessageParser.class.php

```diff
@@ -132,9 +132,15 @@ public function parse($message, $enableSmilies = true, $enableHtml = false, $ena
 	 * @return	string		text
 	 */
 	protected function parseSmilies($text, $enableHtml = false) {
-		foreach ($this->smilies as $code => $html) {
-			//$text = preg_replace('~(?<!&\w{2}|&\w{3}|&\w{4}|&\w{5}|&\w{6}|&#\d{2}|&#\d{3}|&#\d{4}|&#\d{5})'.preg_quote((!$enableHtml ? StringUtil::encodeHTML($code) : $code), '~').'(?![^<]*>)~', $html, $text);
-			$text = preg_replace('~(?<=^|\s|<li>)'.preg_quote((!$enableHtml ? StringUtil::encodeHTML($code) : $code), '~').'(?=$|\s|</li>'.(!$enableHtml ? '|<br />' : '').')~', $html, $text);
+		$codes = array_keys($this->smilies);
+		$pattern = '~(?<=^|\s|<li>)'.str_replace('@', '|', preg_quote((!$enableHtml ? StringUtil::encodeHTML(implode('@', $codes)) : implode('|', $codes)), '~')).'(?=$|\s|</li>'.(!$enableHtml ? '|<br />' : '').')~';
```

Comment:
> get escaped when calling preg_quote

This already does not work properly when `$enableHtml = true`.

`implode('|', array_map('preg_quote', $codes))` would work, if you use the `\wcf\system\Regex` class. `implode('|', array_map(function ($item) { return preg_quote($item, '~'); }))` would work directly with `preg_replace`.